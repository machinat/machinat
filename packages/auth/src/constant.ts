export const STATE_COOKIE_KEY = 'machinat_auth_state';
export const TOKEN_COOKIE_KEY = 'machinat_auth_token';
export const SIGNATURE_COOKIE_KEY = 'machinat_auth_signature';
export const ERROR_COOKIE_KEY = 'machinat_auth_error';

export const TOKEN_STORAGE_KEY = 'machinat_auth_token';
