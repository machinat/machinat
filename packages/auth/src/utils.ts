import type { ServerResponse, IncomingMessage } from 'http';
import { posix as posixPath } from 'path';
import getRawBody from 'raw-body';
import {
  parse as parseCookie,
  serialize as serializeCookie,
  CookieSerializeOptions,
} from 'cookie';
import type {
  WithHeaders,
  AuthApiResponseBody,
  AuthApiErrorBody,
} from './types.js';

export const respondRedirect = (res: ServerResponse, url: string): void => {
  res.writeHead(302, { Location: url });
  res.end();
};

export const getCookies = (req: WithHeaders): null | Record<string, string> => {
  if (!req.headers.cookie) {
    return null;
  }
  return parseCookie(req.headers.cookie);
};

const SET_COOKIE = 'Set-Cookie';

export const setCookie = (
  res: ServerResponse,
  key: string,
  value: string,
  options: CookieSerializeOptions
): void => {
  const cookieDesc = serializeCookie(key, value, options);
  const cookiesAlreadySet = res.getHeader(SET_COOKIE);

  res.setHeader(
    SET_COOKIE,
    Array.isArray(cookiesAlreadySet)
      ? [...cookiesAlreadySet, cookieDesc]
      : typeof cookiesAlreadySet === 'string'
      ? [cookiesAlreadySet, cookieDesc]
      : cookieDesc
  );
};

export const isSubdomain = (domain: string, subdomain: string): boolean =>
  subdomain.endsWith(domain) &&
  (subdomain.length === domain.length ||
    subdomain[subdomain.length - domain.length - 1] === '.');

export const isSubpath = (
  path: null | string,
  subpath: null | string
): boolean =>
  !!(path && subpath) && !posixPath.relative(path, subpath).startsWith('..');

const CONTENT_TYPE_JSON = { 'Content-Type': 'application/json' };

export const respondApiOk = (
  res: ServerResponse,
  platform: string,
  token: string
): void => {
  res.writeHead(200, CONTENT_TYPE_JSON);
  const body: AuthApiResponseBody = { platform, token };
  res.end(JSON.stringify(body));
};

export const respondApiError = (
  res: ServerResponse,
  platform: undefined | string,
  code: number,
  reason: string
): void => {
  res.writeHead(code, CONTENT_TYPE_JSON);
  const body: AuthApiErrorBody = { platform, error: { code, reason } };
  res.end(JSON.stringify(body));
};

export const parseJsonBody = async (req: IncomingMessage): Promise<any> => {
  try {
    const rawBody = await getRawBody(req, { encoding: true });
    const body = JSON.parse(rawBody);

    return typeof body === 'object' ? body : null;
  } catch (err) {
    return null;
  }
};
