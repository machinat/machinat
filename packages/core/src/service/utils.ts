import invariant from 'invariant';
import {
  SOCIABLY_SERVICE_CONTAINER,
  SOCIABLY_SERVICE_PROVIDER,
  SOCIABLY_SERVICE_INTERFACE,
} from '../symbol.js';
import ServiceScope from './scope.js';
import ServiceMaker from './maker.js';
import ProvisionMap from './provisionMap.js';
import type {
  Interfaceable,
  ServiceRequirement,
  ServiceContainer,
  ServiceProvider,
  MaybeContainer,
} from './types.js';

export const isServiceContainer = <T>(
  target: MaybeContainer<T>,
): target is ServiceContainer<T, unknown[]> =>
  typeof target === 'function' &&
  '$$typeof' in target &&
  target.$$typeof === SOCIABLY_SERVICE_CONTAINER;

export const isServiceProvider = (
  target: any,
): target is ServiceProvider<unknown, unknown[]> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  target.$$typeof === SOCIABLY_SERVICE_PROVIDER;

export const maybeInjectContainer = <T>(
  scope: ServiceScope,
  maybeContainer: MaybeContainer<T>,
): T =>
  isServiceContainer(maybeContainer)
    ? scope.injectContainer(maybeContainer)
    : maybeContainer;

export const isInterfaceable = (target: any): target is Interfaceable<any> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  (target.$$typeof === SOCIABLY_SERVICE_INTERFACE ||
    target.$$typeof === SOCIABLY_SERVICE_PROVIDER);

export const createEmptyScope = (): ServiceScope =>
  new ServiceScope(new ServiceMaker(new ProvisionMap()), new Map());

export const polishServiceRequirement = <T>(
  dep: Interfaceable<T> | ServiceRequirement<Interfaceable<T>>,
): ServiceRequirement<Interfaceable<T>> => {
  if (isInterfaceable(dep)) {
    return { require: dep, optional: false };
  }

  invariant(
    isInterfaceable(dep?.require),
    `${(dep as any).name || String(dep)} is not a valid interface`,
  );

  return dep;
};

export const validateLifetime = (lifetime: string): void => {
  invariant(
    lifetime === 'singleton' ||
      lifetime === 'scoped' ||
      lifetime === 'transient',
    `${lifetime} is not valid service lifetime`,
  );
};
