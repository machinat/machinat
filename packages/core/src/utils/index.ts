export { default as compose } from './compose';
export { default as formatNode } from './formatNode';
export { default as getTimeId } from './getTimeId';
export { default as mixin } from './mixin';
export * from './isX';
