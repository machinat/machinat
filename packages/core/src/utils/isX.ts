import type {
  MachinatEmpty,
  MachinatElement,
  FragmentElement,
  PauseElement,
  ThunkElement,
  RawElement,
  ProviderElement,
  FunctionalElement,
  FunctionalComponent,
  ContainerElement,
  ContainerComponent,
  GeneralElement,
  NativeElement,
  NativeComponent,
} from '../types';
import {
  MACHINAT_FRAGMENT_TYPE,
  MACHINAT_ELEMENT_TYPE,
  MACHINAT_PAUSE_TYPE,
  MACHINAT_THUNK_TYPE,
  MACHINAT_RAW_TYPE,
  MACHINAT_PROVIDER_TYPE,
  MACHINAT_NATIVE_TYPE,
  MACHINAT_SERVICE_CONTAINER,
} from '../symbol';

export const isEmpty = (node: unknown): node is MachinatEmpty =>
  typeof node === 'boolean' || node === null || node === undefined;

export const isElement = (node: unknown): node is MachinatElement<any, any> =>
  typeof node === 'object' &&
  node !== null &&
  (node as any).$$typeof === MACHINAT_ELEMENT_TYPE;

export const isFragmentType = (
  node: MachinatElement<any, any>
): node is FragmentElement => node.type === MACHINAT_FRAGMENT_TYPE;

export const isPauseType = (
  node: MachinatElement<any, any>
): node is PauseElement => node.type === MACHINAT_PAUSE_TYPE;

export const isThunkType = (
  node: MachinatElement<any, any>
): node is ThunkElement => node.type === MACHINAT_THUNK_TYPE;

export const isRawType = (
  node: MachinatElement<any, any>
): node is RawElement => node.type === MACHINAT_RAW_TYPE;

export const isProviderType = (
  node: MachinatElement<any, any>
): node is ProviderElement => node.type === MACHINAT_PROVIDER_TYPE;

export const isFunctionalType = (
  node: MachinatElement<any, any>
): node is FunctionalElement<any, FunctionalComponent<any>> =>
  typeof node.type === 'function' &&
  node.type.$$typeof !== MACHINAT_NATIVE_TYPE &&
  node.type.$$typeof !== MACHINAT_SERVICE_CONTAINER;

export const isContainerType = (
  node: MachinatElement<any, any>
): node is ContainerElement<any, ContainerComponent<any>> =>
  typeof node.type === 'function' &&
  node.type.$$typeof === MACHINAT_SERVICE_CONTAINER;

export const isGeneralType = (
  node: MachinatElement<any, any>
): node is GeneralElement => typeof node.type === 'string';

export const isNativeType = <Component extends NativeComponent<any, any>>(
  node: MachinatElement<any, any>
): node is NativeElement<any, Component> =>
  typeof node.type === 'function' &&
  node.type.$$typeof === MACHINAT_NATIVE_TYPE;

export const isElementTypeValid = (
  node: MachinatElement<any, any>
): boolean => {
  const { type } = node;
  return (
    typeof type === 'string' ||
    typeof type === 'function' ||
    type === MACHINAT_FRAGMENT_TYPE ||
    type === MACHINAT_PAUSE_TYPE ||
    type === MACHINAT_PROVIDER_TYPE ||
    type === MACHINAT_THUNK_TYPE ||
    type === MACHINAT_RAW_TYPE
  );
};
