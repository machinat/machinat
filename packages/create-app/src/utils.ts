export const when =
  (shouldOutput: boolean) =>
  (strs: TemplateStringsArray, ...inputs: unknown[]): string => {
    if (!shouldOutput) {
      return '';
    }

    let outputStr = '';

    for (let i = 0; i < strs.length; i += 1) {
      outputStr += strs[i] + (inputs[i] || '');
    }

    return outputStr;
  };

export const polishFileContent = (
  input: null | string | boolean
): null | string | boolean => {
  if (typeof input !== 'string') {
    return input;
  }
  if (input === '') {
    return null;
  }

  let content = input;

  if (content[0] === '\n') {
    content = content.slice(1);
  }
  if (content[content.length - 1] !== '\n') {
    content = `${content}\n`;
  }

  return content;
};

export const getEmptyEnvs = (dotenvContent: string): Map<string, string[]> => {
  const lines = dotenvContent.split('\n');

  const result = new Map();
  let curPlatform = 'default';

  for (const line of lines) {
    if (line.match(/^[A-Z_]+=$/)) {
      const requiredEnv = line.slice(0, -1);

      const envs = result.get(curPlatform);
      if (envs) {
        envs.push(requiredEnv);
      } else {
        result.set(curPlatform, [requiredEnv]);
      }
    } else if (line.match(/^# .+$/)) {
      curPlatform = line.slice(2);
    }
  }

  return result;
};
