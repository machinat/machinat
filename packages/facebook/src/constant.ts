export const FACEBOOK = 'facebook' as const;
export const FB = 'fb' as const;

export const PATH_PERSONAS = 'me/personas' as const;

export const PATH_FEED = 'me/feed' as const;
export const PATH_PHOTOS = 'me/photos' as const;
export const PATH_VIDEOS = 'me/videos' as const;
