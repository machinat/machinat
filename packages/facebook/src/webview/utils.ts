/* eslint-disable import/prefer-default-export */
import type { ContextDetails } from '@sociably/auth';
import FacebookChat from '../Chat';
import FacebookUser from '../User';
import type { FacebookAuthContext, FacebookAuthData } from './types';

export const getAuthContextDetails = (
  data: FacebookAuthData
): ContextDetails<FacebookAuthContext> => {
  const { page: pageId, id } = data;
  return {
    user: new FacebookUser(pageId, id),
    thread: new FacebookChat(pageId, { id }),
    pageId,
  };
};
