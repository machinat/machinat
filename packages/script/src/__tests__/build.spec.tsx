import moxy from '@moxyjs/moxy';
import Sociably, { RenderingChannel } from '@sociably/core';
import { isContainerType } from '@sociably/core/utils/isX';
import ProcessorP from '../processor';
import build from '../build';
import { SOCIABLY_SCRIPT_TYPE } from '../constant';
import {
  IF,
  THEN,
  ELSE_IF,
  ELSE,
  WHILE,
  PROMPT,
  EFFECT,
  LABEL,
  CALL,
  RETURN,
} from '../keyword';

const initVars = moxy(() => ({}));

const channel = { platform: 'test', uid: 'foo.channel' };

const ChildScript = build(
  {
    name: 'ChildScript',
    initVars,
  },
  <>
    {() => <dolore />}
    <PROMPT set={(_, ctx) => ({ x: ctx.x })} key="childPrompt" />
  </>
);

const MyScript = build(
  {
    name: 'MyScript',
    initVars,
  },
  <>
    <LABEL key="start" />
    {() => <b>Lorem</b>}

    <IF condition={() => false}>
      <THEN>
        <WHILE condition={() => true}>
          <LABEL key="first" />
          {() => <i>ipsum</i>}
          <PROMPT key="ask_1" set={(_, ctx) => ({ a: ctx.a })} />
        </WHILE>
      </THEN>
      <ELSE_IF condition={() => true}>
        <LABEL key="second" />
        {() => <dolor />}
        <PROMPT key="ask_2" set={(_, ctx) => ({ b: ctx.b })} />
        <RETURN value={() => 'fooo'} />
      </ELSE_IF>
      <ELSE>
        <LABEL key="third" />
        {() => 'sit amet,'}
        <CALL
          script={ChildScript}
          params={() => ({ foo: 'bar' })}
          goto="childPrompt"
          key="call_1"
        />
      </ELSE>
    </IF>

    <LABEL key="end" />
    <EFFECT set={() => ({ foo: 'bar' })} />
    {() => 'ad minim veniam'}
  </>
);

test('Script object', () => {
  expect(MyScript.name).toBe('MyScript');
  expect(MyScript.$$typeof).toBe(SOCIABLY_SCRIPT_TYPE);

  expect(MyScript.commands).toMatchSnapshot();
  expect(MyScript.initVars).toBe(initVars);
  expect(MyScript.stopPointIndex).toMatchInlineSnapshot(`
    Map {
      "start" => 0,
      "third" => 3,
      "call_1" => 4,
      "first" => 7,
      "ask_1" => 8,
      "second" => 11,
      "ask_2" => 12,
      "end" => 15,
    }
  `);

  expect(ChildScript.name).toBe('ChildScript');
  expect(ChildScript.$$typeof).toBe(SOCIABLY_SCRIPT_TYPE);

  expect(ChildScript.commands).toMatchSnapshot();
  expect(ChildScript.initVars).toBe(initVars);
  expect(ChildScript.stopPointIndex).toMatchInlineSnapshot(`
    Map {
      "childPrompt" => 1,
    }
  `);
});

test('Script.Start', () => {
  const mockProcessor = moxy({
    start: async () => ({ output: () => 'Hello Script' }),
  });

  expect(typeof MyScript.Start).toBe('function');
  expect(isContainerType(<MyScript.Start />)).toBe(true);
  expect(MyScript.Start.$$name).toBe('MyScript');
  expect(MyScript.Start.$$deps).toEqual([
    { require: ProcessorP, optional: false },
    { require: RenderingChannel, optional: false },
  ]);

  expect((MyScript.Start as any)(mockProcessor, channel)({})).resolves.toBe(
    'Hello Script'
  );
  expect(mockProcessor.start.mock).toHaveBeenCalledTimes(1);
  expect(mockProcessor.start.mock).toHaveBeenCalledWith(channel, MyScript, {});

  expect(typeof ChildScript.Start).toBe('function');
  expect(isContainerType(<ChildScript.Start />)).toBe(true);
  expect(ChildScript.Start.$$name).toBe('ChildScript');

  expect(
    (ChildScript.Start as any)(
      mockProcessor,
      channel
    )({
      params: { foo: 'baz' },
      goto: 'bar',
    })
  ).resolves.toBe('Hello Script');
  expect(mockProcessor.start.mock).toHaveBeenCalledTimes(2);
  expect(mockProcessor.start.mock).toHaveBeenCalledWith(channel, ChildScript, {
    params: { foo: 'baz' },
    goto: 'bar',
  });
});
