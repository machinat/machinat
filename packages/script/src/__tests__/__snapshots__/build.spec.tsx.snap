// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Script object 1`] = `
Array [
  Object {
    "getContent": [Function],
    "type": "content",
  },
  Object {
    "condition": [Function],
    "isNot": false,
    "offset": 5,
    "type": "jump_cond",
  },
  Object {
    "condition": [Function],
    "isNot": false,
    "offset": 9,
    "type": "jump_cond",
  },
  Object {
    "getContent": [Function],
    "type": "content",
  },
  Object {
    "goto": "childPrompt",
    "key": "call_1",
    "script": Object {
      "$$typeof": Symbol(lib.script.machinat),
      "Start": [Function],
      "commands": Array [
        Object {
          "getContent": [Function],
          "type": "content",
        },
        Object {
          "key": "childPrompt",
          "setVars": [Function],
          "type": "prompt",
        },
      ],
      "initVars": [Function],
      "name": "ChildScript",
      "stopPointIndex": Map {
        "childPrompt" => 1,
      },
    },
    "setVars": undefined,
    "type": "call",
    "withParams": [Function],
  },
  Object {
    "offset": 10,
    "type": "jump",
  },
  Object {
    "condition": [Function],
    "isNot": true,
    "offset": 4,
    "type": "jump_cond",
  },
  Object {
    "getContent": [Function],
    "type": "content",
  },
  Object {
    "key": "ask_1",
    "setVars": [Function],
    "type": "prompt",
  },
  Object {
    "offset": -3,
    "type": "jump",
  },
  Object {
    "offset": 5,
    "type": "jump",
  },
  Object {
    "getContent": [Function],
    "type": "content",
  },
  Object {
    "key": "ask_2",
    "setVars": [Function],
    "type": "prompt",
  },
  Object {
    "getValue": [Function],
    "type": "return",
  },
  Object {
    "offset": 1,
    "type": "jump",
  },
  Object {
    "setVars": [Function],
    "type": "effect",
    "yieldValue": undefined,
  },
  Object {
    "getContent": [Function],
    "type": "content",
  },
]
`;

exports[`Script object 3`] = `
Array [
  Object {
    "getContent": [Function],
    "type": "content",
  },
  Object {
    "key": "childPrompt",
    "setVars": [Function],
    "type": "prompt",
  },
]
`;
